github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14090083?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14090083?buildTab=artifacts#meta) on refs/heads/master @ [9e60abf53295](https://github.com/cockroachdb/pebble/commits/9e60abf53295):


```
db1.Set("qvkgrjxug", "cqaaqoezpsbccputhsf")
external0 = batch0.NewExternalObj()
db1.Get("hmkq@1")
db1.Get("qvkgrjxug")
db1.DeleteRange("hmkq@1", "qvkgrjxug")
db1.Delete("hmkq@1")
db1.SingleDelete("qvkgrjxug", false /* maybeReplaceDelete */)
db1.LogData("ntruc")
db1.Delete("hmkq@1")
db1.Set("idcbkgvfe@1", "fazcqqbonggict")
batch1 = db1.NewIndexedBatch()
db1.Merge("hmkq@1", "zwuluo")
batch1.Set("basnf@1", "xtu")
db1.Get("idcbkgvfe@1")
db1.Ingest(batch1)
db1.Get("jsbkdnb@1")
db1.Get("idcbkgvfe@1")
db1.Delete("basnf@1")
db1.Set("basnf@1", "ibxbcrzaovin")

--- FAIL: TestMeta (0.00s)
panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1") [recovered]
	panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1") [recovered]
	panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1")

goroutine 5 [running]:
testing.tRunner.func1.2({0xc92880, 0xc000052cc0})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc92880?, 0xc000052cc0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:208 +0xa5
panic({0xc92880?, 0xc000052cc0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0000f8000)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1324 +0x1345
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:212 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc0000ac820, {0x7ffeda25a93c, 0xf}, {0xc0000b8990, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000072760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc0000ac820, 0xdfce78)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/14090085?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/14090085?buildTab=artifacts#meta) on refs/heads/master @ [9e60abf53295](https://github.com/cockroachdb/pebble/commits/9e60abf53295):


```
iter2.SeekLT("bznj@3", "")
db1.Delete("zllbwplytybf@1")
iter2.SeekLT("jqsjffqhs@1", "")
iter4.Next("")
iter3.Prev("")
iter3.Prev("")
iter3.Last()
iter4.SeekGE("uzcuyxu@2", "")
iter3.SeekLT("zllbwp@1", "")
iter4.Prev("")
db1.Delete("edmw")
db1.Get("zclygokdl@7")
iter3.Prev("")
iter3.SetBounds("zllbwpgajugf@2", "zllbwpgajugf@2")
iter3.SeekGE("zllbwpgajugf@2", "")
iter3.Prev("")
iter2.SeekPrefixGE("cpavvwyq@4")
iter4.Last()

panic: could not generate unique key [recovered]
	panic: could not generate unique key [recovered]
	panic: could not generate unique key

goroutine 5 [running]:
testing.tRunner.func1.2({0x11836c0, 0x1433e80})
	GOROOT/src/testing/testing.go:1545 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x716
panic({0x11836c0?, 0x1433e80?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:208 +0xe9
panic({0x11836c0?, 0x1433e80?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).UniqueKeys(0xc000129ba0, 0x2, 0xc000182d00)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:73 +0x34d
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0001538c0)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1333 +0x10ee
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:212 +0x199a
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc000129a00, {0x7fffdf855935, 0xf}, {0xc0001b9a10, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0xbd0
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0x0?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x879
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x29
testing.tRunner(0xc000129a00, 0x12f1770)
	GOROOT/src/testing/testing.go:1595 +0x262
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x846
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14107710?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14107710?buildTab=artifacts#meta) on refs/heads/master @ [e05c22dcae73](https://github.com/cockroachdb/pebble/commits/e05c22dcae73):


```
iter26.SeekPrefixGE("xbajgs@26")
iter24.Next("")
db1.Set("nazolvemctafkdsublk@8", "csdhddyxxaukvecuhwj")
iter25.Prev("")
iter26.Last()
iter23.SeekLT("jacpzn", "")
iter24.SeekPrefixGE("nazolvemcthefqi@14")
snap14.Close()
iter25.Next("shfp")
iter23.SeekLT("qgqwflbcytm", "")
iter26.SeekPrefixGE("xkkrjajots@13")
iter27.SeekLT("nazolvemct@7", "bwrgdhrrhq@7")
iter24.Next("")
db1.Get("brxqbfa@26")
iter27.SeekGE("bwrgdhrrhq", "thqsufzhczq")
db1.Set("nazolvemctajbyaza@17", "zapofdfhxgvajhjemni")
iter26.SeekGE("qyenmqetes", "")
iter25.SeekPrefixGE("nazolvemctajbyaza@19")

panic: could not generate unique key [recovered]
	panic: could not generate unique key [recovered]
	panic: could not generate unique key

goroutine 34 [running]:
testing.tRunner.func1.2({0xc92aa0, 0xebefd0})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc92aa0?, 0xebefd0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:208 +0xa5
panic({0xc92aa0?, 0xebefd0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).UniqueKeys(0xc0002049c0, 0x2, 0xc0001d90c0)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:73 +0x245
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc000276000)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1333 +0x9a5
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:212 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc000204820, {0x7ffe77ebb93c, 0xf}, {0xc0002189c0, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000078760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc000204820, 0xdfd0e8)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14141954?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14141954?buildTab=artifacts#meta) on refs/heads/master @ [0b9461943bcf](https://github.com/cockroachdb/pebble/commits/0b9461943bcf):


```
db1.DeleteRange("xyirfahfqi@1", "ymrpoqrit@1")
db1.Get("ymrpoqrit")
db1.Delete("ljnfg@1")
iter2.SetBounds("eikinhyxsh@2", "icwtt@3")
iter2.SeekGE("eikinhyxsh@2", "")
iter2.Prev("")
iter2.Next("")
db1.Get("fxlnn@3")
db1.Set("qbfcf@4", "rvibarvapkprfcu")
iter2.Last()
db1.Get("eikinhyxsh@2")
iter4.SeekLT("xpqqvahgfli@1", "")
iter5.SeekGE("gabvynpn@2", "")
iter3.SeekGE("fxlnn@3", "")
iter2.Prev("")
db1.SingleDelete("mrtctw@3", false /* maybeReplaceDelete */)

panic: invalid bounds ["pgbk@2", "pgbk@2") [recovered]
	panic: invalid bounds ["pgbk@2", "pgbk@2") [recovered]
	panic: invalid bounds ["pgbk@2", "pgbk@2")

goroutine 7 [running]:
testing.tRunner.func1.2({0xc96cc0, 0xc000053d90})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc96cc0?, 0xc000053d90?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:209 +0xa5
panic({0xc96cc0?, 0xc000053d90?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).randKey(0xc0000cbd40, 0x3f847ae147ae147b, 0xc0001d9038)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:133 +0x1174
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).RandKeyInRange(...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:39
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0000f38c0)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1297 +0x49c
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:213 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc0000cbba0, {0x7ffe15cd793b, 0xf}, {0xc0001df9b0, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000078760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc0000cbba0, 0xe01940)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14193247?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14193247?buildTab=artifacts#meta) on refs/heads/master @ [7531ef4d20c9](https://github.com/cockroachdb/pebble/commits/7531ef4d20c9):


```
iter53.Next("")
iter53.Next("")
iter53.Prev("")
iter48.Last()
iter42.SeekLT("bdcnxgysfb@2", "")
iter43.SeekLT("kopjpfjl@6", "enkotvwno@16")
iter42.First()
iter53.Next("")
snap38.Get("rkdtcitbygs@18")
iter48.SeekPrefixGE("qhatu@17")
db1.Merge("ttjrbz@3", "ganpbmxv")
db1.Set("sifcgnwht@2", "qo")
iter54.Next("kekmflrso")
batch42 = db1.NewIndexedBatch()
db1.Ingest(batch42)
db1.Merge("zrvzpv@13", "sigmy")

panic: invalid randKey "tfofqo@2"; bounds: ["tfofqo@3", "tfofqo@2") 1 0 [recovered]
	panic: invalid randKey "tfofqo@2"; bounds: ["tfofqo@3", "tfofqo@2") 1 0 [recovered]
	panic: invalid randKey "tfofqo@2"; bounds: ["tfofqo@3", "tfofqo@2") 1 0

goroutine 34 [running]:
testing.tRunner.func1.2({0xc9f6c0, 0xc000052070})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc9f6c0?, 0xc000052070?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:209 +0xa5
panic({0xc9f6c0?, 0xc000052070?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).randKey(0xc0002049c0, 0x3f847ae147ae147b, 0xc0001d9008)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:239 +0x926
github.com/cockroachdb/pebble/metamorphic.(*keyGenerator).RandKeyInRange(...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/key_generator.go:39
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc000276000)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1301 +0x568
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:213 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc000204820, {0x7ffc87e7b93c, 0xf}, {0xc0002189c0, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000078760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc000204820, 0xe0ad88)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/ingest.go
+++ b/ingest.go
@@ -1143,7 +1143,7 @@ type ExternalFile struct {
 	//
 	// SyntheticSuffix can only be used under the following conditions:
 	//  - the synthetic suffix must sort before any non-empty suffixes in the
-	//    backing sst;
+	//    backing sst (the entire sst, not just the part restricted to Bounds).
 	//  - the backing sst must not contain multiple keys with the same prefix.
 	SyntheticSuffix []byte
 }

--- a/metamorphic/build.go
+++ b/metamorphic/build.go
@@ -20,9 +20,8 @@ import (
 )
 
 // writeSSTForIngestion writes an SST that is to be ingested, either directly or
-// as an external file.
+// as an external file. Returns the sstable metadata.
 //
-// Returns the sstable metadata and the minimum non-empty suffix.
 // Closes the iterators in all cases.
 func writeSSTForIngestion(
 	t *Test,
@@ -33,7 +32,7 @@ func writeSSTForIngestion(
 	syntheticSuffix sstable.SyntheticSuffix,
 	writable objstorage.Writable,
 	targetFMV pebble.FormatMajorVersion,
-) (_ *sstable.WriterMetadata, minSuffix []byte, _ error) {
+) (*sstable.WriterMetadata, error) {
 	writerOpts := t.opts.MakeWriterOptions(0, targetFMV.MaxTableFormat())
 	if t.testOpts.disableValueBlocksForIngestSSTables {
 		writerOpts.DisableValueBlocks = true
@@ -47,17 +46,10 @@ func writeSSTForIngestion(
 	defer rangeKeyIterCloser.Close()
 
 	outputKey := func(key []byte) []byte {
-		n := t.opts.Comparer.Split(key)
-		if suffix := key[n:]; len(suffix) > 0 {
-			if minSuffix == nil || t.opts.Comparer.Compare(suffix, minSuffix) < 0 {
-				minSuffix = slices.Clone(suffix)
-			}
-		}
 		return slices.Clone(key)
 	}
 
 	if len(syntheticSuffix) > 0 {
-		minSuffix = slices.Clone(syntheticSuffix)
 		outputKey = func(key []byte) []byte {
 			n := t.opts.Comparer.Split(key)
 			return append(key[:n:n], syntheticSuffix...)
@@ -90,30 +82,30 @@ func writeSSTForIngestion(
 		}
 		valBytes, _, err := value.Value(nil)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		k := *key
 		k.UserKey = outputKey(k.UserKey)
 		if err := w.Add(k, valBytes); err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 	}
 	if err := pointIterCloser.Close(); err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	if rangeDelIter != nil {
 		span, err := rangeDelIter.First()
 		for ; span != nil; span, err = rangeDelIter.Next() {
 			if err := w.DeleteRange(outputKey(span.Start), outputKey(span.End)); err != nil {
-				return nil, nil, err
+				return nil, err
 			}
 		}
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		if err := rangeDelIterCloser.Close(); err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 	}
 
@@ -142,25 +134,25 @@ func writeSSTForIngestion(
 			}
 			keyspan.SortKeysByTrailer(&collapsed.Keys)
 			if err := rangekey.Encode(&collapsed, w.AddRangeKey); err != nil {
-				return nil, nil, err
+				return nil, err
 			}
 		}
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		if err := rangeKeyIterCloser.Close(); err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 	}
 
 	if err := w.Close(); err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	sstMeta, err := w.Metadata()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	return sstMeta, minSuffix, nil
+	return sstMeta, nil
 }
 
 // buildForIngest builds a local SST file containing the keys in the given batch
@@ -178,7 +170,7 @@ func buildForIngest(
 	iter, rangeDelIter, rangeKeyIter := private.BatchSort(b)
 
 	writable := objstorageprovider.NewFileWritable(f)
-	meta, _, err := writeSSTForIngestion(
+	meta, err := writeSSTForIngestion(
 		t,
 		iter, rangeDelIter, rangeKeyIter,
 		false, /* uniquePrefixes */
@@ -212,7 +204,7 @@ func buildForIngestExternalEmulation(
 	// emulating the external ingestion path which won't remove duplicate prefixes
 	// if they exist.
 	const uniquePrefixes = false
-	meta, _, err := writeSSTForIngestion(
+	meta, err := writeSSTForIngestion(
 		t,
 		pointIter, rangeDelIter, rangeKeyIter,
 		uniquePrefixes,

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -784,9 +784,9 @@ func (g *generator) iterSeekGEWithLimit(iterID objID) {
 
 // inRangeKeys returns all keys in the range [lower, upper) associated with the
 // given object.
-func (g *generator) inRangeKeys(lower, upper []byte, o objID) []*keyMeta {
-	var inRangeKeys []*keyMeta
-	for _, keyMeta := range g.keyManager.sortedKeysForObj(o) {
+func (g *generator) inRangeKeys(lower, upper []byte, o objID) []keyMeta {
+	var inRangeKeys []keyMeta
+	for _, keyMeta := range g.keyManager.SortedKeysForObj(o) {
 		if g.cmp(keyMeta.key, lower) >= 0 && g.cmp(keyMeta.key, upper) < 0 {
 			inRangeKeys = append(inRangeKeys, keyMeta)
 		}

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -13,7 +13,6 @@ import (
 
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/pebble/internal/randvar"
-	"github.com/cockroachdb/pebble/sstable"
 	"golang.org/x/exp/rand"
 )
 
@@ -1287,6 +1286,13 @@ func (g *generator) writerIngestExternalFiles() {
 	dbID := g.dbs.rand(g.rng)
 	numFiles := 1 + g.expRandInt(1)
 	objs := make([]externalObjWithBounds, numFiles)
+
+	// We generate the parameters in multiple passes:
+	//  1. Generate objs with random start and end keys. Their bounds can overlap.
+	//  2. Sort objects by the start bound and trim the bounds to remove overlap.
+	//  3. Remove any objects where the previous step resulted in empty bounds.
+	//  4. Randomly add synthetic suffixes.
+
 	for i := range objs {
 		// We allow the same object to be selected multiple times.
 		id := g.externalObjects.rand(g.rng)
@@ -1319,22 +1325,16 @@ func (g *generator) writerIngestExternalFiles() {
 		if g.cmp(start, end) == 0 {
 			end = objEnd
 		}
-
-		var syntheticSuffix sstable.SyntheticSuffix
-		if g.rng.Intn(2) == 0 {
-			syntheticSuffix = g.keyGenerator.SkewedSuffix(0.1)
-		}
-
 		objs[i] = externalObjWithBounds{
 			externalObjID: id,
 			bounds: pebble.KeyRange{
 				Start: start,
 				End:   end,
 			},
-			syntheticSuffix: syntheticSuffix,
 		}
 	}
 
+	// Sort by start bound.
 	slices.SortFunc(objs, func(a, b externalObjWithBounds) int {
 		return g.cmp(a.bounds.Start, b.bounds.Start)
 	})
@@ -1350,6 +1350,19 @@ func (g *generator) writerIngestExternalFiles() {
 	objs = slices.DeleteFunc(objs, func(o externalObjWithBounds) bool {
 		return g.cmp(o.bounds.Start, o.bounds.End) >= 0
 	})
+
+	for i := range objs {
+		// We can only use a synthetic suffix if we don't have range dels.
+		// TODO(radu): we will want to support this at some point.
+		if g.keyManager.objKeyMeta(objs[i].externalObjID).hasRangeDels {
+			continue
+		}
+
+		if g.rng.Intn(2) == 0 {
+			// Generate a suffix that sorts before any previously generated suffix.
+			objs[i].syntheticSuffix = g.keyGenerator.IncMaxSuffix()
+		}
+	}
 
 	// The batches we're ingesting may contain single delete tombstones that
 	// when applied to the writer result in nondeterminism in the deleted key.
@@ -1380,6 +1393,10 @@ func (g *generator) writerIngestExternalFiles() {
 		dbID: dbID,
 		objs: objs,
 	})
+}
+
+func (g *generator) keysForExternalIngest(obj externalObjWithBounds) []keyMeta {
+	return g.inRangeKeys(obj.bounds.Start, obj.bounds.End, obj.externalObjID)
 }
 
 func (g *generator) writerLogData() {

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -781,18 +781,6 @@ func (g *generator) iterSeekGEWithLimit(iterID objID) {
 	})
 }
 
-// inRangeKeys returns all keys in the range [lower, upper) associated with the
-// given object.
-func (g *generator) inRangeKeys(lower, upper []byte, o objID) []keyMeta {
-	var inRangeKeys []keyMeta
-	for _, keyMeta := range g.keyManager.SortedKeysForObj(o) {
-		if g.cmp(keyMeta.key, lower) >= 0 && g.cmp(keyMeta.key, upper) < 0 {
-			inRangeKeys = append(inRangeKeys, keyMeta)
-		}
-	}
-	return inRangeKeys
-}
-
 func (g *generator) iterSeekPrefixGE(iterID objID) {
 	lower := g.itersLastOpts[iterID].lower
 	upper := g.itersLastOpts[iterID].upper
@@ -808,7 +796,7 @@ func (g *generator) iterSeekPrefixGE(iterID objID) {
 	// random key.
 	if g.rng.Intn(10) >= 1 {
 		possibleKeys := make([][]byte, 0, 100)
-		inRangeKeys := g.inRangeKeys(lower, upper, g.objDB[iterID])
+		inRangeKeys := g.keyManager.InRangeKeysForObj(g.objDB[iterID], lower, upper)
 		for _, keyMeta := range inRangeKeys {
 			visibleHistory := keyMeta.history.before(iterCreationTimestamp)
 
@@ -1364,20 +1352,20 @@ func (g *generator) writerIngestExternalFiles() {
 		}
 	}
 
-	// The batches we're ingesting may contain single delete tombstones that
-	// when applied to the writer result in nondeterminism in the deleted key.
-	// If that's the case, we can restore determinism by first deleting the keys
-	// from the writer.
+	// The batches we're ingesting may contain single delete tombstones that when
+	// applied to the db result in nondeterminism in the deleted key. If that's
+	// the case, we can restore determinism by first deleting the keys from the
+	// db.
 	//
 	// Generating additional operations here is not ideal, but it simplifies
 	// single delete invariants significantly.
+	dbKeys := g.keyManager.objKeyMeta(dbID)
 	for _, o := range objs {
-		singleDeleteConflicts := g.keyManager.checkForSingleDelConflicts(o.externalObjID, dbID, true /* collapsed */)
-		for _, key := range singleDeleteConflicts {
-			if g.cmp(key, o.bounds.Start) >= 0 && g.cmp(key, o.bounds.End) < 0 {
+		for _, src := range g.keyManager.KeysForExternalIngest(o) {
+			if g.keyManager.checkForSingleDelConflict(src, dbKeys) {
 				g.add(&deleteOp{
 					writerID:    dbID,
-					key:         key,
+					key:         src.key,
 					derivedDBID: dbID,
 				})
 			}
@@ -1393,10 +1381,6 @@ func (g *generator) writerIngestExternalFiles() {
 		dbID: dbID,
 		objs: objs,
 	})
-}
-
-func (g *generator) keysForExternalIngest(obj externalObjWithBounds) []keyMeta {
-	return g.inRangeKeys(obj.bounds.Start, obj.bounds.End, obj.externalObjID)
 }
 
 func (g *generator) writerLogData() {

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -1041,9 +1041,18 @@ func (g *generator) newExternalObj() {
 	if len(g.liveBatches) == 0 {
 		return
 	}
-	batchID := g.liveBatches.rand(g.rng)
-	if g.keyManager.objKeyMeta(batchID).bounds.IsUnset() {
-		return
+	var batchID objID
+	// Try to find a suitable batch.
+	for i := 0; ; i++ {
+		if i == 10 {
+			return
+		}
+		batchID = g.liveBatches.rand(g.rng)
+		okm := g.keyManager.objKeyMeta(batchID)
+		// #3287: IngestExternalFiles currently doesn't support range keys.
+		if !okm.bounds.IsUnset() && !okm.hasRangeKeys {
+			break
+		}
 	}
 	g.removeBatchFromGenerator(batchID)
 	objID := makeObjID(externalObjTag, g.init.externalObjSlots)

--- a/metamorphic/key_generator.go
+++ b/metamorphic/key_generator.go
@@ -102,6 +102,14 @@ func (kg *keyGenerator) SkewedSuffixInt(incMaxProb float64) int64 {
 	return int64(kg.cfg.writeSuffixDist.Uint64(kg.rng))
 }
 
+// IncMaxSuffix increases the max suffix range and returns the new maximum
+// suffix (which is guaranteed to be larger than any previously generated
+// suffix).
+func (kg *keyGenerator) IncMaxSuffix() []byte {
+	kg.cfg.writeSuffixDist.IncMax(1)
+	return testkeys.Suffix(int64(kg.cfg.writeSuffixDist.Max()))
+}
+
 // UniformSuffix returns a suffix in the same range as SkewedSuffix but with a
 // uniform distribution. This is used during reads to better exercise reading a
 // mix of older and newer keys. The suffix can be empty.

--- a/metamorphic/key_manager.go
+++ b/metamorphic/key_manager.go
@@ -265,15 +265,15 @@ func (k *keyManager) objKeyMeta(o objID) *objKeyMeta {
 	return m
 }
 
-// sortedKeysForObj returns all the values in objKeyMeta(o).keys, in sorted
+// SortedKeysForObj returns all the entries in objKeyMeta(o).keys, in sorted
 // order.
-func (k *keyManager) sortedKeysForObj(o objID) []*keyMeta {
+func (k *keyManager) SortedKeysForObj(o objID) []keyMeta {
 	okm := k.objKeyMeta(o)
-	res := make([]*keyMeta, 0, len(okm.keys))
+	res := make([]keyMeta, 0, len(okm.keys))
 	for _, m := range okm.keys {
-		res = append(res, m)
+		res = append(res, *m)
 	}
-	slices.SortFunc(res, func(a, b *keyMeta) int {
+	slices.SortFunc(res, func(a, b keyMeta) int {
 		cmp := k.comparer.Compare(a.key, b.key)
 		if cmp == 0 {
 			panic(fmt.Sprintf("distinct keys %q and %q compared as equal", a.key, b.key))
@@ -395,7 +395,7 @@ func (k *keyManager) doObjectBoundsOverlap(objIDs []objID) bool {
 func (k *keyManager) checkForSingleDelConflicts(srcObj, dstObj objID, srcCollapsed bool) [][]byte {
 	dstKeys := k.objKeyMeta(dstObj)
 	var conflicts [][]byte
-	for _, src := range k.sortedKeysForObj(srcObj) {
+	for _, src := range k.SortedKeysForObj(srcObj) {
 		// Single delete generation logic already ensures that both srcObj and
 		// dstObj's single deletes are deterministic within the context of their
 		// existing writes. However, applying srcObj on top of dstObj may

--- a/metamorphic/key_manager.go
+++ b/metamorphic/key_manager.go
@@ -214,6 +214,9 @@ type objKeyMeta struct {
 	// bounds holds user key bounds encompassing all the keys set within an
 	// object. It's updated within `update` when a new op is generated.
 	bounds bounds
+	// These flags are true if the object has had range del or range key operations.
+	hasRangeDels bool
+	hasRangeKeys bool
 }
 
 // MergeKey adds the given key at the given meta timestamp, merging the histories as needed.
@@ -235,6 +238,18 @@ func (okm *objKeyMeta) CollapseKeys() {
 	for _, keyMeta := range okm.keys {
 		keyMeta.history = keyMeta.history.collapsed()
 	}
+}
+
+// MergeFrom merges the `from` metadata into this one, appending all of its
+// individual operations at the provided timestamp.
+func (okm *objKeyMeta) MergeFrom(from *objKeyMeta, metaTimestamp int, cmp base.Compare) {
+	// The result should be the same regardless of the ordering of the keys.
+	for _, k := range from.keys {
+		okm.MergeKey(k, metaTimestamp)
+	}
+	okm.bounds.Expand(cmp, from.bounds)
+	okm.hasRangeDels = okm.hasRangeDels || from.hasRangeDels
+	okm.hasRangeKeys = okm.hasRangeKeys || from.hasRangeKeys
 }
 
 // objKeyMeta looks up the objKeyMeta for a given object, creating it if necessary.
@@ -328,21 +343,14 @@ func (k *keyManager) getOrInit(id objID, key []byte) *keyMeta {
 	return m
 }
 
-// mergeKeysInto merges the keys and bounds from an object into another. Assumes
-// the source object will not be used again.
-func (k *keyManager) mergeKeysInto(from, to objID) {
-	fromMeta := k.objKeyMeta(from)
+// mergeObjectInto merges obj key metadata from an object into another and
+// deletes the metadata for the source object (which must not be used again).
+func (k *keyManager) mergeObjectInto(from, to objID) {
 	toMeta := k.objKeyMeta(to)
 	ts := k.nextMetaTimestamp()
-	// The result should be the same, regardless of the ordering of the keys.
-	for _, keyMeta := range fromMeta.keys {
-		toMeta.MergeKey(keyMeta, ts)
-	}
-	// All the keys in `from` have been merged into `to`. Expand `to`'s bounds
-	// to be at least as wide as `from`'s.
-	toMeta.bounds.Expand(k.comparer.Compare, fromMeta.bounds)
+	toMeta.MergeFrom(k.objKeyMeta(from), ts, k.comparer.Compare)
 
-	delete(k.byObj, from) // Unlink "from" obj.
+	delete(k.byObj, from)
 }
 
 // expandBounds expands the incrementally maintained bounds of o to be at least
@@ -532,6 +540,8 @@ func (k *keyManager) update(o op) {
 			}
 		}
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
+		k.objKeyMeta(s.writerID).hasRangeDels = true
+
 	case *singleDeleteOp:
 		meta := k.getOrInit(s.writerID, s.key)
 		meta.history = append(meta.history, keyHistoryItem{
@@ -546,10 +556,13 @@ func (k *keyManager) update(o op) {
 		// of those batches are, as that determines whether that ingestion
 		// will succeed or not.
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
+		k.objKeyMeta(s.writerID).hasRangeKeys = true
 	case *rangeKeySetOp:
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
+		k.objKeyMeta(s.writerID).hasRangeKeys = true
 	case *rangeKeyUnsetOp:
 		k.expandBounds(s.writerID, k.makeEndExclusiveBounds(s.start, s.end))
+		k.objKeyMeta(s.writerID).hasRangeKeys = true
 	case *ingestOp:
 		// Some ingestion operations may attempt to ingest overlapping sstables
 		// which is prohibited. We know at generation time whether these
@@ -568,7 +581,7 @@ func (k *keyManager) update(o op) {
 		// merge that.
 		for _, batchID := range s.batchIDs {
 			k.objKeyMeta(batchID).CollapseKeys()
-			k.mergeKeysInto(batchID, s.dbID)
+			k.mergeObjectInto(batchID, s.dbID)
 		}
 		// TODO(bilal): Handle ingestAndExciseOp and replicateOp here. We currently
 		// disable SingleDelete when these operations are enabled (see
@@ -576,7 +589,7 @@ func (k *keyManager) update(o op) {
 	case *newExternalObjOp:
 		// Collapse and transfer the keys from the batch to the external object.
 		k.objKeyMeta(s.batchID).CollapseKeys()
-		k.mergeKeysInto(s.batchID, s.externalObjID)
+		k.mergeObjectInto(s.batchID, s.externalObjID)
 	case *ingestExternalFilesOp:
 		// Merge the keys from the external objects (within the restricted bounds)
 		// into the database.
@@ -595,10 +608,10 @@ func (k *keyManager) update(o op) {
 		}
 	case *applyOp:
 		// Merge the keys from this batch into the parent writer.
-		k.mergeKeysInto(s.batchID, s.writerID)
+		k.mergeObjectInto(s.batchID, s.writerID)
 	case *batchCommitOp:
 		// Merge the keys from the batch with the keys from the DB.
-		k.mergeKeysInto(s.batchID, s.dbID)
+		k.mergeObjectInto(s.batchID, s.dbID)
 	}
 }
 

--- a/metamorphic/key_manager.go
+++ b/metamorphic/key_manager.go
@@ -283,6 +283,32 @@ func (k *keyManager) SortedKeysForObj(o objID) []keyMeta {
 	return res
 }
 
+// InRangeKeysForObj returns all keys in the range [lower, upper) associated with the
+// given object, in sorted order.
+func (k *keyManager) InRangeKeysForObj(o objID, lower, upper []byte) []keyMeta {
+	var inRangeKeys []keyMeta
+	for _, km := range k.SortedKeysForObj(o) {
+		if k.comparer.Compare(km.key, lower) >= 0 && k.comparer.Compare(km.key, upper) < 0 {
+			inRangeKeys = append(inRangeKeys, km)
+		}
+	}
+	return inRangeKeys
+
+}
+
+// KeysForExternalIngest returns the keys that will be ingested with an external
+// object (taking into consideration the bounds, synthetic suffix, etc).
+func (k *keyManager) KeysForExternalIngest(obj externalObjWithBounds) []keyMeta {
+	keys := k.InRangeKeysForObj(obj.externalObjID, obj.bounds.Start, obj.bounds.End)
+	if obj.syntheticSuffix.IsSet() {
+		for i := range keys {
+			n := k.comparer.Split(keys[i].key)
+			keys[i].key = append(keys[i].key[:n:n], obj.syntheticSuffix...)
+		}
+	}
+	return keys
+}
+
 func (k *keyManager) nextMetaTimestamp() int {
 	ret := k.metaTimestamp
 	k.metaTimestamp++
@@ -396,102 +422,115 @@ func (k *keyManager) checkForSingleDelConflicts(srcObj, dstObj objID, srcCollaps
 	dstKeys := k.objKeyMeta(dstObj)
 	var conflicts [][]byte
 	for _, src := range k.SortedKeysForObj(srcObj) {
-		// Single delete generation logic already ensures that both srcObj and
-		// dstObj's single deletes are deterministic within the context of their
-		// existing writes. However, applying srcObj on top of dstObj may
-		// violate the invariants. Consider:
-		//
-		//    src: a.SET; a.SINGLEDEL;
-		//    dst: a.SET;
-		//
-		// The merged view is:
-		//
-		//    a.SET; a.SET; a.SINGLEDEL;
-		//
-		// This is invalid, because there is more than 1 value mutation of the
-		// key before the single delete.
-		//
-		// We walk the source object's history in chronological order, looking
-		// for a single delete that was written before a DEL/RANGEDEL. (NB: We
-		// don't need to look beyond a DEL/RANGEDEL, because these deletes bound
-		// any subsequently-written single deletes to applying to the keys
-		// within src's history between the two tombstones. We already know from
-		// per-object history invariants that any such single delete must be
-		// deterministic with respect to src's keys.)
-		var srcHasUnboundedSingleDelete bool
-		var srcValuesBeforeSingleDelete int
-
-		// When the srcObj is being ingested (srcCollapsed=t), the semantics
-		// change. We must first "collapse" the key's history to represent the
-		// ingestion semantics.
-		srcHistory := src.history
 		if srcCollapsed {
-			srcHistory = src.history.collapsed()
+			src.history = src.history.collapsed()
 		}
-
-	srcloop:
-		for _, item := range srcHistory {
-			switch item.opType {
-			case OpWriterDelete, OpWriterDeleteRange:
-				// We found a DEL or RANGEDEL before any single delete. If src
-				// contains additional single deletes, their effects are limited
-				// to applying to later keys. Combining the two object histories
-				// doesn't pose any determinism risk.
-				break srcloop
-			case OpWriterSingleDelete:
-				// We found a single delete. Since we found this single delete
-				// before a DEL or RANGEDEL, this delete has the potential to
-				// affect the visibility of keys in `dstObj`. We'll need to look
-				// for potential conflicts down below.
-				srcHasUnboundedSingleDelete = true
-				if srcValuesBeforeSingleDelete > 1 {
-					panic(errors.AssertionFailedf("unexpectedly found %d sets/merges within %s before single del",
-						srcValuesBeforeSingleDelete, srcObj))
-				}
-				break srcloop
-			case OpWriterSet, OpWriterMerge:
-				// We found a SET or MERGE operation for this key. If there's a
-				// subsequent single delete, we'll need to make sure there's not
-				// a SET or MERGE in the dst too.
-				srcValuesBeforeSingleDelete++
-			default:
-				panic(errors.AssertionFailedf("unexpected optype %d", item.opType))
-			}
-		}
-		if !srcHasUnboundedSingleDelete {
-			continue
-		}
-
-		dst, ok := dstKeys.keys[string(src.key)]
-		// If the destination writer has no record of the key, the combined key
-		// history is simply the src object's key history which is valid due to
-		// per-object single deletion invariants.
-		if !ok {
-			continue
-		}
-
-		// We need to examine the trailing key history on dst.
-		consecutiveValues := srcValuesBeforeSingleDelete
-	dstloop:
-		for i := len(dst.history) - 1; i >= 0; i-- {
-			switch dst.history[i].opType {
-			case OpWriterSet, OpWriterMerge:
-				// A SET/MERGE may conflict if there's more than 1 consecutive
-				// SET/MERGEs.
-				consecutiveValues++
-				if consecutiveValues > 1 {
-					conflicts = append(conflicts, src.key)
-					break dstloop
-				}
-			case OpWriterDelete, OpWriterSingleDelete, OpWriterDeleteRange:
-				// Dels clear the history, enabling use of single delete.
-				break dstloop
-			default:
-				panic(errors.AssertionFailedf("unexpected optype %d", dst.history[i].opType))
-			}
+		if k.checkForSingleDelConflict(src, dstKeys) {
+			conflicts = append(conflicts, src.key)
 		}
 	}
 	return conflicts
+}
+
+// checkForSingleDelConflict returns true if applying the history of the source
+// key on top of the given object results in a possible SingleDel
+// nondeterminism. See checkForSingleDelConflicts.
+func (k *keyManager) checkForSingleDelConflict(src keyMeta, dstObjKeyMeta *objKeyMeta) bool {
+	// Single delete generation logic already ensures that both the source
+	// object and the destination object's single deletes are deterministic
+	// within the context of their existing writes. However, applying the source
+	// keys on top of the destination object may violate the invariants.
+	// Consider:
+	//
+	//    src: a.SET; a.SINGLEDEL;
+	//    dst: a.SET;
+	//
+	// The merged view is:
+	//
+	//    a.SET; a.SET; a.SINGLEDEL;
+	//
+	// This is invalid, because there is more than 1 value mutation of the
+	// key before the single delete.
+	//
+	// We walk the source object's history in chronological order, looking
+	// for a single delete that was written before a DEL/RANGEDEL. (NB: We
+	// don't need to look beyond a DEL/RANGEDEL, because these deletes bound
+	// any subsequently-written single deletes to applying to the keys
+	// within src's history between the two tombstones. We already know from
+	// per-object history invariants that any such single delete must be
+	// deterministic with respect to src's keys.)
+	var srcHasUnboundedSingleDelete bool
+	var srcValuesBeforeSingleDelete int
+
+	// When the srcObj is being ingested (srcCollapsed=t), the semantics
+	// change. We must first "collapse" the key's history to represent the
+	// ingestion semantics.
+	srcHistory := src.history
+
+srcloop:
+	for _, item := range srcHistory {
+		switch item.opType {
+		case OpWriterDelete, OpWriterDeleteRange:
+			// We found a DEL or RANGEDEL before any single delete. If src
+			// contains additional single deletes, their effects are limited
+			// to applying to later keys. Combining the two object histories
+			// doesn't pose any determinism risk.
+			return false
+
+		case OpWriterSingleDelete:
+			// We found a single delete. Since we found this single delete
+			// before a DEL or RANGEDEL, this delete has the potential to
+			// affect the visibility of keys in `dstObj`. We'll need to look
+			// for potential conflicts down below.
+			srcHasUnboundedSingleDelete = true
+			if srcValuesBeforeSingleDelete > 1 {
+				panic(errors.AssertionFailedf("unexpectedly found %d sets/merges before single del",
+					srcValuesBeforeSingleDelete))
+			}
+			break srcloop
+
+		case OpWriterSet, OpWriterMerge:
+			// We found a SET or MERGE operation for this key. If there's a
+			// subsequent single delete, we'll need to make sure there's not
+			// a SET or MERGE in the dst too.
+			srcValuesBeforeSingleDelete++
+
+		default:
+			panic(errors.AssertionFailedf("unexpected optype %d", item.opType))
+		}
+	}
+	if !srcHasUnboundedSingleDelete {
+		return false
+	}
+
+	dst, ok := dstObjKeyMeta.keys[string(src.key)]
+	// If the destination writer has no record of the key, the combined key
+	// history is simply the src object's key history which is valid due to
+	// per-object single deletion invariants.
+	if !ok {
+		return false
+	}
+
+	// We need to examine the trailing key history on dst.
+	consecutiveValues := srcValuesBeforeSingleDelete
+	for i := len(dst.history) - 1; i >= 0; i-- {
+		switch dst.history[i].opType {
+		case OpWriterSet, OpWriterMerge:
+			// A SET/MERGE may conflict if there's more than 1 consecutive
+			// SET/MERGEs.
+			consecutiveValues++
+			if consecutiveValues > 1 {
+				return true
+			}
+		case OpWriterDelete, OpWriterSingleDelete, OpWriterDeleteRange:
+			// Dels clear the history, enabling use of single delete.
+			return false
+
+		default:
+			panic(errors.AssertionFailedf("unexpected optype %d", dst.history[i].opType))
+		}
+	}
+	return false
 }
 
 // update updates the internal state of the keyManager according to the given
@@ -596,13 +635,8 @@ func (k *keyManager) update(o op) {
 		ts := k.nextMetaTimestamp()
 		dbMeta := k.objKeyMeta(s.dbID)
 		for _, obj := range s.objs {
-			for _, keyMeta := range k.objKeyMeta(obj.externalObjID).keys {
-				if k.comparer.Compare(obj.bounds.Start, keyMeta.key) <= 0 &&
-					k.comparer.Compare(keyMeta.key, obj.bounds.End) < 0 {
-					// Note: the keys have already been collapsed when the external object
-					// was created.
-					dbMeta.MergeKey(keyMeta, ts)
-				}
+			for _, keyMeta := range k.KeysForExternalIngest(obj) {
+				dbMeta.MergeKey(&keyMeta, ts)
 			}
 			dbMeta.bounds.Expand(k.comparer.Compare, k.makeEndExclusiveBounds(obj.bounds.Start, obj.bounds.End))
 		}

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -934,7 +934,7 @@ type externalObjWithBounds struct {
 func (o *ingestExternalFilesOp) run(t *Test, h historyRecorder) {
 	db := t.getDB(o.dbID)
 
-	// We modify objs to eliminate empty objects and clear invalid synthetic suffixes.
+	// We modify objs to eliminate empty objects.
 	var objs []externalObjWithBounds
 	for _, obj := range o.objs {
 		// Make sure the object exists and is not empty.
@@ -948,18 +948,6 @@ func (o *ingestExternalFilesOp) run(t *Test, h historyRecorder) {
 			// TODO(radu): even though we don't expect this case in practice, eventually
 			// we want to make sure that it doesn't cause failures.
 			continue
-		}
-		if obj.syntheticSuffix != nil {
-			// Verify that the version supports suffix replacement, and verify that
-			// the suffix comes before any suffix present in the object.
-			if t.opts.Comparer.Compare(obj.syntheticSuffix, objMeta.minSuffix) >= 0 {
-				obj.syntheticSuffix = nil
-			}
-			if objMeta.sstMeta.HasRangeDelKeys {
-				// Disable synthetic suffix if we have range dels.
-				// TODO(radu): we will want to support this at some point.
-				obj.syntheticSuffix = nil
-			}
 		}
 		objs = append(objs, obj)
 	}
@@ -1770,7 +1758,7 @@ func (o *newExternalObjOp) run(t *Test, h historyRecorder) {
 
 	iter, rangeDelIter, rangeKeyIter := private.BatchSort(b)
 
-	sstMeta, minSuffix, err := writeSSTForIngestion(
+	sstMeta, err := writeSSTForIngestion(
 		t,
 		iter, rangeDelIter, rangeKeyIter,
 		true, /* uniquePrefixes */
@@ -1787,8 +1775,7 @@ func (o *newExternalObjOp) run(t *Test, h historyRecorder) {
 		panic("external object has range keys")
 	}
 	t.setExternalObj(o.externalObjID, externalObjMeta{
-		sstMeta:   sstMeta,
-		minSuffix: minSuffix,
+		sstMeta: sstMeta,
 	})
 	h.Recordf("%s", o)
 }

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -1770,12 +1770,6 @@ func (o *newExternalObjOp) run(t *Test, h historyRecorder) {
 
 	iter, rangeDelIter, rangeKeyIter := private.BatchSort(b)
 
-	// #3287: range keys are not supported yet.
-	if rangeKeyIter != nil {
-		rangeKeyIter.Close()
-		rangeKeyIter = nil
-	}
-
 	sstMeta, minSuffix, err := writeSSTForIngestion(
 		t,
 		iter, rangeDelIter, rangeKeyIter,
@@ -1786,6 +1780,11 @@ func (o *newExternalObjOp) run(t *Test, h historyRecorder) {
 	)
 	if err != nil {
 		panic(err)
+	}
+	if sstMeta.HasRangeKeys {
+		// #3287: IngestExternalFiles currently doesn't support range keys; we check
+		// for range keys in newExternalObj.
+		panic("external object has range keys")
 	}
 	t.setExternalObj(o.externalObjID, externalObjMeta{
 		sstMeta:   sstMeta,

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -1007,7 +1007,7 @@ func (o *ingestExternalFilesOp) syncObjs() objIDSlice {
 func (o *ingestExternalFilesOp) String() string {
 	strs := make([]string, len(o.objs))
 	for i, obj := range o.objs {
-		strs[i] = fmt.Sprintf("%s, %q, %q, %q", obj.externalObjID, obj.bounds.Start, obj.bounds.End, obj.syntheticSuffix)
+		strs[i] = fmt.Sprintf("%s, %q, %q, %q /* syntheticSuffix */", obj.externalObjID, obj.bounds.Start, obj.bounds.End, obj.syntheticSuffix)
 	}
 	return fmt.Sprintf("%s.IngestExternalFiles(%s)", o.dbID, strings.Join(strs, ", "))
 }

--- a/metamorphic/test.go
+++ b/metamorphic/test.go
@@ -69,9 +69,6 @@ type Test struct {
 
 type externalObjMeta struct {
 	sstMeta *sstable.WriterMetadata
-	// minSuffix is the minimum (according to the comparator) non-empty suffix in
-	// the object.
-	minSuffix []byte
 }
 
 func newTest(ops []op) *Test {

